cpp/src/arrow/python/numpy_to_arrow.cc

```diff
@@ -164,18 +163,26 @@ static Status AppendObjectBinaries(PyArrayObject* arr, PyArrayObject* mask,
     if ((have_mask && mask_values[offset]) || PandasObjectIsNull(obj)) {
       RETURN_NOT_OK(builder->AppendNull());
       continue;
-    } else if (!PyBytes_Check(obj)) {
+    } else if (PyBytes_Check(obj)) {
```

Comment:
We're probably doing this kind of dance (taking a bytes object and extracting a pointer and size) in other places already. I think it would be nice to factor that that somewhere (see e.g. `src/arrow/python/helpers.h`). That would also allow supporting bytearray in other places.